Sources/AardvarkReveal/ARKRevealArchiveBuilder.m

```diff
@@ -0,0 +1,285 @@
+        /* mode */      {'0', '0', '0', '7', '7', '7', ' ', '\0'},
+        /* uid */       {'0', '0', '0', '0', '0', '0', ' ', '\0'},
+        /* gid */       {'0', '0', '0', '0', '0', '0', ' ', '\0'},
+        /* size */      {'0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', ' '},
+        /* mtime */     {'0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', ' '},
+        /* checksum */  {' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '},
+        /* typeflag */  type,
+        /* linkname */  "",
+        /* magic */     TMAGIC,
+        /* version */   TVERSION,
+        /* uname */     {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+        /* gname */     {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+        /* devmajor */  {'0', '0', '0', '0', '0', '0', ' ', '\0'},
+        /* devminor */  {'0', '0', '0', '0', '0', '0', ' ', '\0'},
+        /* prefix */    "",
+        /* pad */       ""
+    };
+
+    // Per the ustar spec, we should check the length of the file name and split it across the `name` and `prefix`
+    // fields if it's too long. In practice, we have a very shallow file hierarchy and short file names, so we should
+    // never overflow past the `name` field.
+
+    const char *_Nullable pathString = [path cStringUsingEncoding:NSUTF8StringEncoding];
+    memcpy(&header.name, pathString, [path lengthOfBytesUsingEncoding:NSUTF8StringEncoding]);
+
+    if (linkName != nil) {
+        const char *_Nullable linkString = [linkName cStringUsingEncoding:NSUTF8StringEncoding];
+        memcpy(&header.linkname, linkString, [linkName lengthOfBytesUsingEncoding:NSUTF8StringEncoding]);
+    }
+
+    int failed = 0;
+
+    failed |= write_ascii_octal(fileSize, (char *)(&header.size), sizeof(header.size) - 1);
+
+    failed |= write_ascii_octal(_modificationDate, (char *)(&header.mtime), sizeof(header.mtime) - 1);
+
+    int64_t checksum = 0;
+    for (int i = 0 ; i < TAR_BLOCK_LENGTH ; i++) {
+        checksum += 255 & (int64_t)(((char *)&header)[i]);
+    }
+    failed |= write_ascii_octal(checksum, (char *)&header.checksum, 6);
+
+    if (failed) {
+        *error = [NSError errorWithDomain:ARKRevealArchiveBuilderErrorDomain
+                                     code:ARKRevealArchiveBuilderErrorCodeHeaderWriteFailure
+                                 userInfo:nil];
+    }
+
+    return header;
+}
+
+/// Writes the value `v` as an ASCII representation of an octal number of length `s` to `p`.
+///
+/// Returns `0` if successful, otherwise `1`.
+static int write_ascii_octal(int64_t v, char *p, int s) {
+    // Octal values can't be negative.
+    if (v < 0) {
+        return 1;
+    }
+
+    // Start at the least significant digit and work backwards. Fill the entire length of the string, including leading
+    // zeros.
+    p += s;
+    while (s-- > 0) {
+        *--p = (char)('0' + (v & 7));
+        v >>= 3;
```

Comment:
what are these magic numbers? the 7 and 3? Can we make them named constants?

Reply:
This is the conversion to octal (base 8). Doing a bitwise and with 7 gets you the last three bits (`7` in decimal is `111` in binary). Add that to `'0'` and you get that number in ASCII (`'0'` through `'7'`). Then you shift by 3 bits to get the next least significant digit.

Maybe it would be clearer to replace `7` with `0b111`?

Reply:
`0b111` would be a bit more clear. I'd also strongly recommend wring tests for this code.

Reply:
I'm just a fly on the wall, but I think this could also be made clearer by taking the explanation you've provided here on GitHub and turning that into more granular comments (rather than one comment at the start of this code block). That way, you're walking the reader through it a bit more, and readers will have to brush up less on their bit manipulation 🙏🏻

Reply:
Addressed in 462da40 and aa78c12